### PR TITLE
Demonstrating a simple UX test pattern to validate the XAML and UIKit properties using the FunctionalTest framework

### PR DESCRIPTION
**Pull request by oliversa-msft**
Demonstrating a simple UX test pattern to validate the XAML and UIKit properties using the FunctionalTest framework

1. CALayerAppearanceTests.mm - 2 tests - validating opacity and backgroundColor between the UIKit and XAML layers
2. UIButtonsTests.mm - 1 test - re-use XAMLCatalog's buttonViewController to validate the state of a button backgroundColor property

CHANGES:
=======
1. Moved all UIKit Tests under its own folder/filter
2. Created a CoreAnimation test suite for this iteration's CALayer tests

TODO:
====
1. ARC-enabled these UX tests where possible
2. Formalize the relationship between UIButton tests that reach into existing VCs (e.g. XAMLCatalog)
3. Jared doesn't like NSCondition - explore if there are other options that work better :-)

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1483)

### CODE REVIEW

File: tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    self->_layer.frame = self->_viewForLayer.bounds;
+    [self->_viewForLayer.layer addSublayer:self->_layer];
+}
+@end
+
+//
+// Validate that the CALayer opacity property change invokes a property change in the backing XAML element
+//
+TEST(CALayerAppearance, OpacityChanged) {
+    __block CALayerViewController* caLayerVC;
```
Comment (yiyang-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: >__block [](start = 4, length = 7)

do these have to be __block?

File: tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    __block NSCondition* condition = [[[NSCondition alloc] init] autorelease];
+
+    dispatch_sync(dispatch_get_main_queue(), ^{
+        LOG_INFO("Creating CALayerViewController on the UI thread explicitly");
+        caLayerVC = [[CALayerViewController alloc] init];
+        [caLayerVC view];
+
+        // We have to artificially ref the element since the block needs to keep it around
+        WXUIElement* backingElement = [caLayerVC.layer _xamlElement];
+        [backingElement retain];
```
Comment (yiyang-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: >backingElement [](start = 9, length = 14)

so if callback didn't fire for some reason, are you leaking this?

Comment (oliversa-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: Hmmmm, in the failure case they would probably leak before its own object and the entire UIApplication is torn down. We could tackle it with ARC which is what I'm investigating as the next step.

---
In reply to: [90351052](https://github.com/Microsoft/WinObjC/pull/1483#discussion_r90351052) [](ancestors = 90351052)

Comment (yiyang-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: may as well move this outside of dispatch and use __block WXUIElement* backingElement =nil;
and do creation and retain as usual
and release at the end after check if backingElement is not nil. or better 
    auto reset = wil::ScopeExit([]() {
        if(backingElement) {
             [backingElement release]
        }
    });


---
In reply to: [90351052](https://github.com/Microsoft/WinObjC/pull/1483#discussion_r90351052) [](ancestors = 90351052)

Comment (oliversa-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: Nice! Unfortunately, the scopexit doesn't like __block variables but we can still explicitly release it for now.

---
In reply to: [90351820](https://github.com/Microsoft/WinObjC/pull/1483#discussion_r90351820) [](ancestors = 90351820,90351052)

**Comment by yiyang-msft** [comment]
:shipit:

**Comment by jaredhms** [comment]
:shipit:

File: tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+// Validate that the CALayer opacity property change invokes a property change in the backing XAML element
+//
+TEST(CALayerAppearance, OpacityChanged) {
+    __block CALayerViewController* caLayerVC;
+    __block NSCondition* condition = [[[NSCondition alloc] init] autorelease];
+
+    dispatch_sync(dispatch_get_main_queue(), ^{
+        LOG_INFO("Creating CALayerViewController on the UI thread explicitly");
+        caLayerVC = [[CALayerViewController alloc] init];
+        [caLayerVC view];
```
Comment (oliversa-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: No problem 

---
In reply to: [90349626](https://github.com/Microsoft/WinObjC/pull/1483#discussion_r90349626) [](ancestors = 90349626)

File: tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        // We have to artificially ref the element since the block needs to keep it around
+        WXUIElement* backingElement = [caLayerVC.layer _xamlElement];
+        [backingElement retain];
+
+        // Register callback and wait for the property changed event to trigger
+        __block int64_t callbackToken = [backingElement
```
Comment (oliversa-msft) [inline tests/functionaltests/Tests/CoreAnimationTests/CALayerAppearanceTests.mm]: No, it's only used to unregister. I guess it's safe to remove __block.

---
In reply to: [90350182](https://github.com/Microsoft/WinObjC/pull/1483#discussion_r90350182) [](ancestors = 90350182)

**Pull request merged**